### ISSUE TITLE
Empty matches for `search_around_sky` returns (empty) floating point arrays

### ISSUE THREAD

**Issue by evertrol** [astropy/astropy#4877]
When no match at all in `search_around_sky` can be made, the index arrays returned are of floating point dtype:

``` python
>>> from astropy.coordinates import SkyCoord, search_around_sky
>>> from astropy import units
>>> c1 = SkyCoord(ra=[0], dec=[0], unit=units.deg)
>>> c2 = SkyCoord(ra=[180], dec=[0], unit=units.deg)
>>> index1, index2, _, _ = search_around_sky(c1, c2, 1*units.deg)
>>> index1.dtype, index2.dtype
(dtype('float64'), dtype('float64'))
>>> len(index1), len(index2)
(0, 0)
```

Obviously, it's straightforward to test for the length of the output indices, but since an empty array can be used to index a `SkyCoord`, it'be nice to have `search_around_sky` return empty index arrays with the correct type in these cases.

A simply fix might be:

```
diff --git a/astropy/coordinates/matching.py b/astropy/coordinates/matching.py
index ee22888..8887981 100644
--- a/astropy/coordinates/matching.py
+++ b/astropy/coordinates/matching.py
@@ -354,8 +354,8 @@ def search_around_sky(coords1, coords2, seplimit, storekdtree='_kdtree_sky'):
         for match in matches:
             idxs1.append(i)
             idxs2.append(match)
-    idxs1 = np.array(idxs1)
-    idxs2 = np.array(idxs2)
+    idxs1 = np.array(idxs1, dtype=np.int64)
+    idxs2 = np.array(idxs2, dtype=np.int64)

     if idxs1.size == 0:
         d2ds = u.Quantity([], u.deg)
```

(Tangentially related: #4875 )

**Comment by pllim** [comment]
Does it need to be explicitly `int64`? Anyways, this seems reasonable to me, so feel free to open a PR. Thanks!

**Comment by evertrol** [comment]
`int64` just happens to be what `numpy.int_` is on my system: essentially a `C long`. `np.int`, `np.int_` or even simply `int` is better.
I'll check whether the other matching routines (e.g. `match_coordinates_sky` may have the same issue, add some tests and put in a PR.

Is it worth combining this issue and #4875 in one PR (as both cases should return empty index arrays of correct type)?
Or does #4875 require a bit more thought how to handle that (e.g., raise a ValueError instead, though I prefer returning an empty array)?

**Comment by pllim** [comment]
Not sure... Perhaps @eteq can chime in.

**Comment by evertrol** [comment]
Closed by #5083
